app/src/main/java/com/paranoidandroid/journey/wizard/fragments/LegsFragment.java

```diff
@@ -1,39 +1,138 @@
+import com.paranoidandroid.journey.wizard.adapters.LegsArrayAdapter;
+import com.paranoidandroid.journey.wizard.models.LegItem;
+import com.paranoidandroid.journey.wizard.utils.JourneyBuilder;
+import com.paranoidandroid.journey.wizard.utils.PlaceJSONParser;
+import com.parse.GetCallback;
+import com.parse.ParseException;
+import com.parse.ParseQuery;
+
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import cz.msebera.android.httpclient.Header;
 
 /**
  * Created by epushkarskaya on 11/12/16.
  */
 
 public class LegsFragment extends WizardFragment {
 
-    ListView lvLegs;
-    LegItemArrayAdapter adapter;
+    RecyclerView rvLegs;
+    LegsArrayAdapter adapter;
+    AutoCompleteTextView atvPlaces;
+
+    public static LegsFragment newInstance(String journeyId) {
+        LegsFragment fragment = new LegsFragment();
+
+        Bundle args = new Bundle();
+        args.putString("journey_id", journeyId);
+        fragment.setArguments(args);
+
+        return fragment;
+    }
 
     @Override
     public View onCreateView(LayoutInflater inflater, ViewGroup parent, Bundle savedInstanceState) {
         View v = inflater.inflate(R.layout.fragment_wizard_legs, parent, false);
-        lvLegs = (ListView) v.findViewById(R.id.lvLegs);
-        adapter = new LegItemArrayAdapter(getContext(), listener);
-        lvLegs.setAdapter(adapter);
+        rvLegs = (RecyclerView) v.findViewById(R.id.rvLegs);
+
+        if (getArguments() != null) {
+            String journeyId = getArguments().getString("journey_id");
+            if (journeyId != null) {
+                loadJourneyLegs(journeyId);
+            }
+        } else {
+            adapter = new LegsArrayAdapter(getContext(), new ArrayList<Leg>(), listener);
+        }
+
+        rvLegs.setAdapter(adapter);
+        rvLegs.setLayoutManager(new LinearLayoutManager(getContext()));
 
-        // We want to start with a single item so that users can input first destination
-        adapter.createEmptyRow();
+        atvPlaces = (AutoCompleteTextView) v.findViewById(R.id.atvPlaces);
+        atvPlaces.setThreshold(1);
+
+        atvPlaces.addTextChangedListener(new TextWatcher() {
+
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+               getPlaces(s.toString(), atvPlaces);
+            }
+
+            @Override
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+                // nothing
+            }
+
+            @Override
+            public void afterTextChanged(Editable s) {
+                // nothing
+            }
+        });
 
         return v;
     }
 
-    @Override
-    public boolean readyToPublish() {
-        return adapter.hasValidData();
+    private void getPlaces(String text, final AutoCompleteTextView input){
+        GooglePlaceSearchClient.autoComplete(text, new JsonHttpResponseHandler(){
```

Comment:
This is good for now, but we may need to display spinners in case the network is slow.

Also, I found that the Google Places API provides 3 different ways of looking up places on your behalf (https://developers.google.com/places/android-api/). We should talk about this later.